### PR TITLE
Change plugins directory to spyplugins and make it a namespace package

### PR DESCRIPTION
**Pull request by goanpeca**
This works builds on top of the work made by @Nodd (dont worry, your commits are all there ;-) )

~~I still have an issue when running the translation scripts~~

## Todo
* [x] Change plugins folder from `spyderplugins` to `spyplugins/ui`
* [x] Change plugins folder from `spyderplugins` to `spyplugins/io`
* [x] Fix translations
* [x] Remove `p_` prefix from plugins
* [x] Fix image display in preferences for plugins (pylint and profiler)

### CODE REVIEW

**Comment by goanpeca** [comment]
@ccordoba12 when running `python gettext_scan` I am getting this:

```bash
Updating... spyderlib/locale/fr/LC_MESSAGES/spyderlib.po
merge...
spyderlib/locale/spyderlib.pot:2690: duplicate message definition...
spyderlib/locale/spyderlib.pot:6: ...this is the location of the first definition
msgmerge: found 1 fatal error
Updating... spyderlib/locale/es/LC_MESSAGES/spyderlib.po
merge...
spyderlib/locale/spyderlib.pot:2690: duplicate message definition...
spyderlib/locale/spyderlib.pot:6: ...this is the location of the first definition
msgmerge: found 1 fatal error
Updating... spyderlib/locale/pt_BR/LC_MESSAGES/spyderlib.po
merge...
spyderlib/locale/spyderlib.pot:2690: duplicate message definition...
spyderlib/locale/spyderlib.pot:6: ...this is the location of the first definition
msgmerge: found 1 fatal error
```

**Comment by ccordoba12** [comment]
No idea, I'll check this out after beta1

**Comment by goanpeca** [comment]
@ccordoba12 ... ok this was a bit longer than expected but I think it is now working. I need some testers...

**Comment by ccordoba12** [comment]
I'll test this weekend :-)

**Comment by goanpeca** [comment]
@SylvainCorlay how should we handle icons in external plugins? I mean, we need to define some sane/easy way for people to define both a fontawesome icon and a normal image icon...

**Comment by ccordoba12** [comment]
This needs a rebase too.

Also, I'm getting this traceback in Python 3 (Python 2 works just fine ;-):

```python-traceback
ERROR: 3rd party plugin import failed for `spyplugins.io.hdf5`
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/otherplugins.py", line 53, in _import_plugin
    submodule = imp.load_module(submodule_name, *info)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 171, in load_source
    module = methods.load()
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1444, in exec_module
  File "<frozen importlib._bootstrap>", line 1554, in get_code
  File "<frozen importlib._bootstrap>", line 672, in _code_to_bytecode
ValueError: unmarshallable object
ERROR: 3rd party plugin import failed for `spyplugins.io.dicom`
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/otherplugins.py", line 53, in _import_plugin
    submodule = imp.load_module(submodule_name, *info)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 171, in load_source
    module = methods.load()
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1444, in exec_module
  File "<frozen importlib._bootstrap>", line 1554, in get_code
  File "<frozen importlib._bootstrap>", line 672, in _code_to_bytecode
ValueError: unmarshallable object
ERROR: 3rd party plugin import failed for `spyplugins.io.hdf5`
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/otherplugins.py", line 53, in _import_plugin
    submodule = imp.load_module(submodule_name, *info)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 171, in load_source
    module = methods.load()
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1444, in exec_module
  File "<frozen importlib._bootstrap>", line 1554, in get_code
  File "<frozen importlib._bootstrap>", line 672, in _code_to_bytecode
ValueError: unmarshallable object
ERROR: 3rd party plugin import failed for `spyplugins.io.dicom`
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/otherplugins.py", line 53, in _import_plugin
    submodule = imp.load_module(submodule_name, *info)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/home/carlos/.virtualenvs/test-ipy3-stable/lib/python3.4/imp.py", line 171, in load_source
    module = methods.load()
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1444, in exec_module
  File "<frozen importlib._bootstrap>", line 1554, in get_code
  File "<frozen importlib._bootstrap>", line 672, in _code_to_bytecode
ValueError: unmarshallable object
```

This is printed in both the internal console and also the terminal from which Spyder was started.

@goanpeca, can you reproduce it?

**Comment by goanpeca** [comment]
@ccordoba12 will try to work on this tomorrow

**Comment by goanpeca** [comment]
@ccordoba12 so I think this one is fixed now and should be working on Py2.7 3.3 and 3.4

Can you test it ?

**Comment by goanpeca** [comment]
@ccordoba12 this will not work unless we use setuptools.... distutils cannot handle namespace packages...

**Comment by goanpeca** [comment]
@ccordoba12 now it is ready, can you test?

File: setup.py [hunk 1]
```diff
@@ -244,8 +244,13 @@ def get_packages():
       platforms=['any'],
       packages=get_packages(),
       package_data={LIBNAME: get_package_data(LIBNAME, EXTLIST),
-                    'spyderplugins':
-                    get_package_data('spyderplugins', EXTLIST)},
+                    'spyplugins':
+                    get_package_data('spyplugins', EXTLIST+['.py']),
```
Comment (ccordoba12) [inline setup.py]: Isn't `.py` part of EXTLIST?

Comment (ccordoba12) [inline setup.py]: Another thing: doesn't `get_package_data` works recursively? It would seem like that given that before it was only necessary to pass `spyderplugins` to it :-)

File: spyderlib/config/base.py [hunk 2]
```diff
@@ -196,10 +196,6 @@ def add_image_path(path):
 
 add_image_path(get_module_data_path('spyderlib', relpath='images'))
 
-from spyderlib.otherplugins import PLUGIN_PATH
-if PLUGIN_PATH is not None:
-    add_image_path(osp.join(PLUGIN_PATH, 'images'))
-
```
Comment (ccordoba12) [inline spyderlib/config/base.py]: Why is this not necessary anymore?

File: spyderlib/otherplugins.py [hunk 3]
```diff
@@ -8,52 +8,115 @@
...
+# Declare as a namespace package
+__import__('pkg_resources').declare_namespace(__name__)
+"""
+
+
+def _get_spyderplugins(plugin_path, base_namespace, plugins_namespace,
+                       modnames, modlist):
+    """Scan the directory `plugin_path` for plugins_namespace package and
+    loads its submodules."""
+    namespace_path = Path(plugin_path) / base_namespace / plugins_namespace
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: What's this? I mean, I don't understand how `namespace_path` is built.

Can't you use `os.path` instead for this? I think it'd be more readable :-)

Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Ok, it seems easier to work with, but at least add a comment describing what's saved in `namespace_path`.

File: spyderlib/otherplugins.py [hunk 4]
```diff
@@ -8,52 +8,115 @@
...
+
+    for name in namespace_path.files(pattern="*.py"):
+        if name.name == "__init__.py":
+            continue
+        _import_plugin(name.namebase, base_namespace, plugins_namespace,
+                       namespace_path, modnames, modlist)
+
+
+class _ModuleMock():
+    """ """
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: ???

File: setup.py (hunk 1 above)
Comment (goanpeca) [inline setup.py]: Probably :-)

File: spyderlib/otherplugins.py [hunk 5]
```diff
@@ -8,52 +8,115 @@
...
+    # Import parent module
+    importlib.import_module(namespace)
+
+    # Create user directory
+    user_conf_path = Path(get_conf_path())
+    user_plugin_basepath = user_conf_path / base_namespace
+    user_plugin_path = user_conf_path / base_namespace / plugins_namespace
+    user_plugin_path.makedirs_p()
+    (user_plugin_basepath / "__init__.py").write_text(INIT_PY)
+    (user_plugin_path / "__init__.py").write_text(INIT_PY)
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: What happens with this code if this structure already exists inside `~/.spyder`?

File: spyderlib/config/base.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/config/base.py]: That is from @Nodd

I guess because now the plugins path is not a fixed thing but can be located in many places (namespace packages)

File: spyderlib/otherplugins.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: This was code from Joseph, I work from there, but I will modify this code to make it clearer as soon as it is merged cause I have a plugin manager already setup for this (to enable or disable plugins on startup)

File: spyderlib/plugins/__init__.py [hunk 6]
```diff
@@ -455,6 +462,13 @@ def set_plugin_font(self, font, option=None):
         """Set plugin font option"""
         set_font(font, self.CONF_SECTION, option)
 
+#    def get_icon(self, icon_name):
+#        """
+#        Search icon.
+#        THe icon is searched first in the plugin images directory, then in
+#        spyders's IMG_PATH.
+#        """
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Please remove this lines

File: spyderlib/otherplugins.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: Oops I have to add the docstring....

It was bit tricky and the only way I could get around this... so

When we are loading a plugin it will try to load the translations, but since the get_translations methods depends on a module already loaded in `sys.modules` it was failing, so basically this class is just a mock I use to store a class variable `LOCALEPATH` that get transtalations will use to correctlyfin the files.. once the module loads, I restore the actual module object into `sys.modules`

File: spyderlib/otherplugins.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: @Nodd ....

I will check, but it either overwrites or does nothing (both will yield the same results....)

File: spyderlib/plugins/__init__.py (hunk 6 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Will do

File: spyderlib/utils/icon_manager.py [hunk 7]
```diff
@@ -224,19 +232,25 @@ def get_icon(name, default=None, resample=False):
...
+            qta.load_font('spyder', 'spyder.ttf', 'spyder-charmap.json',
+                          directory=_resource['directory'])
             _resource['loaded'] = True
         args, kwargs = _qtaargs[name]
         return qta.icon(*args, **kwargs)
     elif theme == 'spyder 2':
         icon = get_icon(name + '.png', resample=resample)
+        if icon_path:
+            icon_path = Path(icon_path) / (name + '.png')
+            print(icon_path)
```
Comment (ccordoba12) [inline spyderlib/utils/icon_manager.py]: Remove this `print`

File: spyplugins/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__import__('pkg_resources').declare_namespace(__name__)
```
Comment (ccordoba12) [inline spyplugins/__init__.py]: Please add a docstring here to describe how the new `spyplugins` work :-)

File: spyplugins/io/__init__.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#from pkgutil import extend_path
```
Comment (ccordoba12) [inline spyplugins/io/__init__.py]: Please also add a docstring here

File: spyplugins/ui/__init__.py [hunk 10]
```diff
@@ -0,0 +1 @@
+__import__('pkg_resources').declare_namespace(__name__)
```
Comment (ccordoba12) [inline spyplugins/ui/__init__.py]: Docstring here too please ;-)

File: spyderlib/utils/icon_manager.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/utils/icon_manager.py]: Oops... will do :-)

File: spyplugins/ui/__init__.py (hunk 10 above)
Comment (goanpeca) [inline spyplugins/ui/__init__.py]: Sure.. I will put some links (for the interested reader) also

File: setup.py (hunk 1 above)
Comment (ccordoba12) [inline setup.py]: Ok, please check that out before merging

File: spyderlib/config/base.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/config/base.py]: No problem here, it has to be done now by every plugin now, according to the code below ;-)

File: spyderlib/otherplugins.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Ok

File: spyderlib/otherplugins.py (hunk 4 above)
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Ok, no prob. Just add a docstring describing its purpose.

File: spyderlib/otherplugins.py (hunk 5 above)
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Writes or overwrites? Please check it out before merging.

**Comment by ccordoba12** [comment]
Ok, first review done. It's working without errors on Python 3, but I still have to:

1. Test this on Windows.
2. Run `pylint` and `profiler` with example files to see if they are working as before :-)

Thanks a lot for this @goanpeca and @Nodd. It seems a lot of work, I mean, from the design and implementation point of view.

File: spyplugins/ui/__init__.py (hunk 10 above)
Comment (ccordoba12) [inline spyplugins/ui/__init__.py]: Great!! Thanks a lot :-)

**Comment by goanpeca** [comment]
@ccordoba12 I organized the code a bit better and addressed your questions.

File: spyplugins/__init__.py (hunk 8 above)
Comment (goanpeca) [inline spyplugins/__init__.py]: done

File: spyplugins/ui/__init__.py (hunk 10 above)
Comment (goanpeca) [inline spyplugins/ui/__init__.py]: done

**Comment by ccordoba12** [comment]
Ok, thanks. I'll give another review later today, and also test this on Windows.

**Comment by goanpeca** [comment]
@ccordoba12 have you had time to review and test this?

File: spyderlib/otherplugins.py [hunk 11]
```diff
@@ -8,52 +8,140 @@
...
+        modlist.append(module)
+        modnames.append(module_name)
+    except Exception:
+        sys.stderr.write("ERROR: 3rd party plugin import failed for "
+                         "`{0}`\n".format(module_name))
+        traceback.print_exc(file=sys.stderr)
+
+
+def create_userplugins_files(path):
+    """
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Missing docstring here :-)

File: spyderlib/otherplugins.py [hunk 12]
```diff
@@ -8,52 +8,140 @@
...
+    user_plugin_basepath = user_conf_path / base_namespace
+    user_plugin_path = user_conf_path / base_namespace / plugins_namespace
+
+    create_userplugins_files(user_plugin_basepath)
+    create_userplugins_files(user_plugin_path)
+
+    modlist, modnames = [], []
+
+    # The user plugins directory is given the priority when looking for modules
+    for plugin_path in [user_conf_path] + sys.path:
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: I don't think we need to scan the entire `sys.path` for *our* plugins. Outside `~/.spyder` we know they are going to be located in `spyplugins`.

I think this fact can be easily added to `_get_spyderplugins`

File: spyderlib/py3compat.py [hunk 13]
```diff
@@ -24,7 +24,8 @@
 
 PY2 = sys.version[0] == '2'
 PY3 = sys.version[0] == '3'
-
+PY34 = sys.version_info[0:2] >= (3, 4)
```
Comment (ccordoba12) [inline spyderlib/py3compat.py]: This is not needed anywhere, so please remove it :-)

Comment (goanpeca) [inline spyderlib/py3compat.py]: Ok

File: spyderlib/otherplugins.py (hunk 12 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: We need to load the plugins inside the repo when running in bootstrap mode--

Hmmm @Nodd, help!

File: spyderlib/otherplugins.py (hunk 11 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: ok

File: spyderlib/otherplugins.py [hunk 14]
```diff
@@ -8,52 +8,140 @@
...
+            continue
+        _import_plugin(name.namebase, base_namespace, plugins_namespace,
+                       namespace_path, modnames, modlist)
+
+
+class _ModuleMock():
+    """This mock module is added to sys.modules on plugin load to add the
+    location of the LOCALEDATA so that the module loads succesfully.
+    Once loaded the module is replaced by the actual loaded module object.
+    """
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Can't we add a `pass` here? Just to put some content on it, and also show that does nothing :-)

File: spyderlib/otherplugins.py (hunk 12 above)
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Ok, let's leave as it is now, and later we can see if we can improve it ;-)

File: spyderlib/plugins/__init__.py [hunk 15]
```diff
@@ -264,6 +270,7 @@ def __init__(self, main=None, **kwds):
         """Bind widget to a QMainWindow instance"""
         super(SpyderPluginMixin, self).__init__(**kwds)
         assert self.CONF_SECTION is not None
+        self.PLUGIN_PATH = Path(inspect.getfile(self.__class__)).parent
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: This seems too complex to my taste, so i don't like it too much.

Couldn't we simplify it somehow?

File: spyderlib/utils/icon_manager.py [hunk 16]
```diff
@@ -181,10 +183,16 @@
...
     'attribute':               [('spyder.circle-letter-a',), {'color': 'magenta'}],
     'module':                  [('spyder.circle-letter-m',), {'color': '#daa520'}],
     'no_match':                [('fa.circle',), {'color': 'gray'}],
+    'no_match':                [('fa.circle',), {'color': 'gray'}],
+    # --- Third party plugins ------------------------------------------------
+    'profiler':                [('fa.clock-o',), {}],
+    'pylint':                  [('fa.search', 'fa.check'), {'options': [{}, {'offset': (0.125, 0.125), 'color': 'orange'}]}],
+    'condapackages':           [('fa.archive',), {}],
+    'spyplugins.ui.example':   [('fa.eye',), {}],
+    'spyplugins.ui.autopep8':  [('fa.eye',), {}],
```
Comment (ccordoba12) [inline spyderlib/utils/icon_manager.py]: Two things:

1. How are we going to handle this icon assignment for plugins written by people who are not Spyder developers?
2. Please rename the last two plugins in this list with the names we previously agreed :-) (in their repos and here too).

File: spyderlib/utils/icon_manager.py [hunk 17]
```diff
@@ -1,8 +1,11 @@
 import os
```
Comment (ccordoba12) [inline spyderlib/utils/icon_manager.py]: Copyright assignment is missing in this file. Please add it before we forget about it.

File: spyderlib/utils/icon_manager.py (hunk 16 above)
Comment (goanpeca) [inline spyderlib/utils/icon_manager.py]: I still have not figured out this part.... we need to allow the extension of the qtawesome icons for third party plugins in an easy way. I think we need @SylvainCorlay on this

**Comment by ccordoba12** [comment]
Ok, good news is that this is working without problems on Windows.

Bad news is that I really hate the use of `Path`. I'd prefer to use `osp` as usual (except where it's really not possible) to maintain a single standard (else things could become a mess).

**Comment by ccordoba12** [comment]
@goanpeca, I'm going to do a PR against your branch to fix most of the things I reported today.

I plan to have it ready in one hour :-)

**Comment by goanpeca** [comment]
I agree, although Path can be handy it is a problem to not have uniformity, I will revert to the use of osp.

**Comment by goanpeca** [comment]
Are you sure?, I can do the fixes now as well... there are a couple of PRs open just for your delight :)

**Comment by ccordoba12** [comment]
Ok, ok :-p I just want to merge this ASAP. Then be my guest and do my proposed changes while I review other PRs ;-)

**Comment by goanpeca** [comment]
Ok, will start now ;-)

File: spyderlib/plugins/__init__.py (hunk 15 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: I made all changes (plus a small update to how io plugins load) except this part...

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Actually we no longer need that...!

File: spyplugins/ui/profiler/profiler.py [hunk 18]
```diff
@@ -69,8 +72,9 @@ def get_plugin_title(self):
 
     def get_plugin_icon(self):
         """Return widget icon"""
-        return ima.icon('profiler')
-    
+        path = Path(self.PLUGIN_PATH) / self.IMG_PATH
```
Comment (ccordoba12) [inline spyplugins/ui/profiler/profiler.py]: Please fix this: `self.PLUGIN_PATH` does not exist anymore :-) Its absence generates the following traceback when you try to open Preferences:

```python-traceback
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/spyder.py", line 2663, in edit_preferences
    dlg.add_page(widget)
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/plugins/configdialog.py", line 228, in add_page
    item.setIcon(widget.get_icon())
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/plugins/__init__.py", line 58, in get_icon
    return self.plugin.get_plugin_icon()
  File "/home/carlos/Projects/spyder/github-repo/spyplugins/ui/pylint/pylint.py", line 99, in get_plugin_icon
    path = Path(self.PLUGIN_PATH) / self.IMG_PATH
AttributeError: 'Pylint' object has no attribute 'PLUGIN_PATH'
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/spyder.py", line 2663, in edit_preferences
    dlg.add_page(widget)
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/plugins/configdialog.py", line 228, in add_page
    item.setIcon(widget.get_icon())
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/plugins/__init__.py", line 58, in get_icon
    return self.plugin.get_plugin_icon()
  File "/home/carlos/Projects/spyder/github-repo/spyplugins/ui/profiler/profiler.py", line 75, in get_plugin_icon
    path = Path(self.PLUGIN_PATH) / self.IMG_PATH
AttributeError: 'Profiler' object has no attribute 'PLUGIN_PATH'
```

File: spyplugins/ui/pylint/pylint.py [hunk 19]
```diff
@@ -90,11 +93,12 @@ def __init__(self, parent=None):
     def get_plugin_title(self):
         """Return widget title"""
         return _("Static code analysis")
-    
+
     def get_plugin_icon(self):
         """Return widget icon"""
-        return ima.icon('pylint')
-    
+        path = Path(self.PLUGIN_PATH) / self.IMG_PATH
```
Comment (ccordoba12) [inline spyplugins/ui/pylint/pylint.py]: Also fix this one :-)

File: spyplugins/ui/profiler/profiler.py (hunk 18 above)
Comment (goanpeca) [inline spyplugins/ui/profiler/profiler.py]: Oops, my bad hehe

File: spyderlib/otherplugins.py (hunk 12 above)
Comment (Nodd) [inline spyderlib/otherplugins.py]: They could be installed anywhere on python's path theoretically. For example by using the `--user` option from pip it would be in `~/.local`

**Comment by Nodd** [comment]
I really love Path and hate osp, but I agree that consistency is more important. :cry:

**Comment by goanpeca** [comment]
@nod He is alive!!!

**Comment by ccordoba12** [comment]
This needs a rebase. @goanpeca, I can help you fix the minor problem with IMG_PATH after you do the rebase :-)

**Comment by goanpeca** [comment]
Ok Carlos, you are lucky today, I actually have time to work on this, but in a couple of hours. I will let you know ok :-) ?

**Comment by ccordoba12** [comment]
Great!

**Comment by goanpeca** [comment]
@ccordoba12 sorry I was absent, but I could not work the day as said I would and I was super busy with the real job(s) :wink:

I think I fixed the problems plus made the rebase, need to check what is going on with the continuous integration though

**Comment by goanpeca** [comment]
I had to remove the spyplugins tests, cause they do not run standalone, they need spyder to be running...

File: continuous_integration/travis/modules_test.sh [hunk 20]
```diff
@@ -94,12 +94,3 @@ for f in spyderlib/*/*/*.py; do
         exit 1
     fi
 done
-
-
-# Spyderplugins
-for f in spyderplugins/widgets/*.py; do
```
Comment (ccordoba12) [inline continuous_integration/travis/modules_test.sh]: Please don't remove these tests. Just change this line to

    for f in spyplugins/ui/*/widgets/*.py; do

That should fix these tests because spyplugins widgets can be run outside Spyder.

Comment (goanpeca) [inline continuous_integration/travis/modules_test.sh]: Nope, that does not cut it... it was not only that

Test wont run cause they do not run as standalone scripts, they only work when a spyder instance runs

**Comment by goanpeca** [comment]
@ccordoba12 done,made the change.

For the moment this should do.

But I will think of a better solution to avoid the hack, in the next plugins PR.

All travis tests are passing except 
https://travis-ci.org/spyder-ide/spyder/jobs/89153216

Can you give me a hand?

**Comment by ccordoba12** [comment]
Ok, I'll take a look at it.

**Comment by ccordoba12** [comment]
So, tests in Appveyor are passing (which is terrific!) although we're not testing there the widgets now present in `spyplugins`. I think we should leave that for another PR because it could be a bit tricky :-)

Tomorrow I'll do my final review and a bit of manual testing too ;-)

File: setup.py [hunk 21]
```diff
@@ -274,8 +274,9 @@ def run(self):
       platforms=['any'],
       packages=get_packages(),
       package_data={LIBNAME: get_package_data(LIBNAME, EXTLIST),
-                    'spyderplugins':
-                    get_package_data('spyderplugins', EXTLIST)},
+                    'spyplugins': get_package_data('spyplugins', EXTLIST),
+                    },
+      include_package_data=True,
```
Comment (ccordoba12) [inline setup.py]: Why is this line necessary?

Comment (ccordoba12) [inline setup.py]: i mean the `include_package_data` one :-)

File: spyderlib/otherplugins.py [hunk 22]
```diff
@@ -8,52 +8,151 @@
...
+    pass
+
+
+def _import_plugin(name, base_namespace, plugin_namespace, namespace_path,
+                   modnames, modlist):
+    """Import the plugin `plugins_namsepace`.`name`, add it to `modlist` and
+    adds its name to `modname`."""
```
Comment (ccordoba12) [inline spyderlib/otherplugins.py]: Here is missing an `s` in `modname` , right?

File: spyderlib/plugins/__init__.py [hunk 23]
```diff
@@ -26,6 +26,11 @@
 from spyderlib.qt.QtCore import Qt, Signal, QObject, QEvent, QPoint
 import spyderlib.utils.icon_manager as ima
 
+# Stdlib imports
+import inspect
+import os
+import sys
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: `sys` is not used anymore. Please remove its import

File: spyderlib/plugins/__init__.py [hunk 24]
```diff
@@ -26,6 +26,11 @@
 from spyderlib.qt.QtCore import Qt, Signal, QObject, QEvent, QPoint
 import spyderlib.utils.icon_manager as ima
 
+# Stdlib imports
+import inspect
+import os
+import sys
+
 # Local imports
 from spyderlib.utils.qthelpers import toggle_actions, get_icon, create_action
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: `get_icon` is not here. Please remove it too.

File: spyderlib/spyder.py [hunk 25]
```diff
@@ -1132,13 +1132,14 @@ def add_xydoc(text, pathlist):
...
-            for mod in get_spyderplugins_mods(prefix='p_', extension='.py'):
+            for mod in get_spyderplugins_mods():
                 try:
                     plugin = mod.PLUGIN_CLASS(self)
                     self.thirdparty_plugins.append(plugin)
                     plugin.register_plugin()
-                except AttributeError as error:
+                except Exception as error:
                     print("%s: %s" % (mod, str(error)), file=STDERR)
+                    traceback.print_exc(file=STDERR)
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Why this new line here?

File: spyderlib/utils/icon_manager.py [hunk 26]
```diff
@@ -1,8 +1,16 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2015 The Spyder Development Team
```
Comment (ccordoba12) [inline spyderlib/utils/icon_manager.py]: This line must read

    Copyright © 2009- The Spyder Development Team

Please change it

File: spyderlib/spyder.py (hunk 25 above)
Comment (Nodd) [inline spyderlib/spyder.py]: I think it's to have both the plugin name and the full traceback printed. Both are very useful to plugin or core developers to get informations from users even outside debugging phase so I vote to keep both of them.

File: setup.py (hunk 21 above)
Comment (goanpeca) [inline setup.py]: I think the wheels was not working without but I tried several things so that might not be needed at all :-)

File: spyderlib/otherplugins.py (hunk 22 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: Sure

File: spyderlib/plugins/__init__.py (hunk 23 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: ok

File: spyderlib/plugins/__init__.py (hunk 24 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: ok

File: spyderlib/utils/icon_manager.py (hunk 26 above)
Comment (goanpeca) [inline spyderlib/utils/icon_manager.py]: Should we do it more lie
```bash
Copyright © 2013- The Spyder Development Team
Copyright © 2009-2013 Pierre
```
?

File: setup.py (hunk 21 above)
Comment (ccordoba12) [inline setup.py]: Ok, please see if it's needed or not by running:

    python setup.py bdist_wheel --universal

and checking if `spyplugins` is part of the wheel or not.

File: spyderlib/spyder.py (hunk 25 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: +1 then to keep them both.

File: spyderlib/utils/icon_manager.py (hunk 26 above)
Comment (ccordoba12) [inline spyderlib/utils/icon_manager.py]: Pierre is part of the Team :-) So no, it's not a lie.

I've been changing all our copyrights like the line I mentioned above.

Comment (goanpeca) [inline spyderlib/utils/icon_manager.py]: I dont think that is how it is supposed to work, @PierreRaybaut?

File: setup.py (hunk 21 above)
Comment (goanpeca) [inline setup.py]: Not needed, just checked

File: spyderlib/otherplugins.py (hunk 22 above)
Comment (goanpeca) [inline spyderlib/otherplugins.py]: Fixed

File: spyderlib/plugins/__init__.py (hunk 23 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Done

File: spyderlib/plugins/__init__.py (hunk 24 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Done

File: spyderlib/utils/icon_manager.py (hunk 26 above)
Comment (goanpeca) [inline spyderlib/utils/icon_manager.py]: Done

**Comment by goanpeca** [comment]
@ccordoba12, done with the comments

**Comment by ccordoba12** [comment]
This needs a rebase again, sorry :-)

**Comment by goanpeca** [comment]
Ok, one sec....

**Comment by ccordoba12** [comment]
Ok, it seems the time has come (finally!) to press the green button on this one :-)

Thanks a lot @Nodd and @goanpeca for your work. It's a really great improvement!

**Comment by goanpeca** [comment]
:+1:, now I will push the plugin manager PR :-)

**Comment by goanpeca** [comment]
Side note: conda-manager plugin is a bit broken right now....

**Comment by ccordoba12** [comment]
Ok, no biggie. There are the other plugins to test :-)

**Comment by Nodd** [comment]
Cool, thanks to @goanpeca for finishing the work and @ccordoba12 for the review !
